bigquery/storage/managedwriter/managed_stream.go

```diff
@@ -272,36 +285,53 @@ func (ms *ManagedStream) append(pw *pendingWrite, opts ...gax.CallOption) error
 			req = reqCopy
 		})
 
-		var err error
-		if req == nil {
-			err = (*arc).Send(pw.request)
-		} else {
-			// we had to amend the initial request
+		// critical section:  When we issue an append, we need to add the write to the pending channel
+		// to keep the response ordering correct.
+		ms.mu.Lock()
+		if req != nil {
+			// First append in a new connection needs properties like schema and stream name set.
 			err = (*arc).Send(req)
-		}
-		recordStat(ms.ctx, AppendRequests, 1)
-		recordStat(ms.ctx, AppendRequestBytes, int64(pw.reqSize))
-		recordStat(ms.ctx, AppendRequestRows, int64(len(pw.request.GetProtoRows().Rows.GetSerializedRows())))
-		if err != nil {
-			status := grpcstatus.Convert(err)
-			if status != nil {
-				ctx, _ := tag.New(ms.ctx, tag.Insert(keyError, status.Code().String()))
-				recordStat(ctx, AppendRequestErrors, 1)
-			}
-			bo, shouldRetry := r.Retry(err)
-			if shouldRetry {
-				if err := gax.Sleep(ms.ctx, bo); err != nil {
-					return err
-				}
-				continue
-			}
-			ms.mu.Lock()
-			ms.err = err
-			ms.mu.Unlock()
+		} else {
+			// Subsequent requests need no modification.
+			err = (*arc).Send(pw.request)
 		}
 		if err == nil {
+			// Compute numRows, once we pass ownership to the channel the request may be
+			// cleared.
+			numRows := int64(len(pw.request.GetProtoRows().Rows.GetSerializedRows()))
 			ch <- pw
```

Comment:
Should this guard against the `otherCtx` expiring as well?

Something like 

```go
select {
    case ch <- pw:
        // We've passed ownership of the pending write to the channel.
        // It's now responsible for marking the request done, we're done
        // with the critical section.
        ms.mu.Unlock()

        // Record stats and return.
        recordStat(ms.ctx, AppendRequests, 1)
        recordStat(ms.ctx, AppendRequestBytes, int64(pw.reqSize))
        recordStat(ms.ctx, AppendRequestRows, numRows)
        return nil
  case <-otherCtx.Done():
       ms.mu.Unlock()
       return otherCtx.Err()
}
```

That way the call to `append` is guaranteed not to lock if `ch` is full and it is guaranteed to return within the given context timeout (within some margins).

And once calls to `append` are guaranteed to return within the timeout the goroutine in `AppendRows` could probably be removed and replaced with a synchronous call I believe.

Reply:
The issue is that the select allows the writer to get out of sync processing append responses, which arrive in the same order as append requests were accepted.  We use the channel for maintaining the ordering when processing the responses.